## Read proMGE database
- source: https://promge.embl.de/about.cgi 

In [1]:
import pandas as pd 
from pandarallel import pandarallel

# Initialization
pandarallel.initialize(progress_bar=True)

import warnings 
warnings.filterwarnings("ignore") 


INFO: Pandarallel will run on 12 workers.
INFO: Pandarallel will use standard multiprocessing data transfer (pipe) to transfer data between the main process and workers.


In [2]:
promge_df = pd.read_csv("dataset/prev_studies/proMGE/mges.txt", sep="\t")

promge_df['taxid'] = promge_df.apply(lambda x: int(x["mge_genome_position"].split(".")[0]), axis=1)
promge_df['biosample_id'] = promge_df.apply(lambda x: x["mge_genome_position"].split(".")[1], axis=1)
promge_df['position'] = promge_df.apply(lambda x: x["mge_genome_position"].split(".")[-1], axis=1)
promge_df['Accession'] = promge_df['position'].str.extract(r'^([^:]+)')
# Apply extract to create two new columns
promge_df[['Start', 'End']] = promge_df['position'].str.extract(r':(\d+)-(\d+)').astype(int)

In [3]:
promge_df.head()

,mge_genome_position,mge_length,number_of_proteins,phage_genes,conjugation_genes,recombinase,genome,mge_category,count,specI,...,order,family,genus,species,taxid,biosample_id,position,Accession,Start,End
0,562.SAMN04376786.LVNY01000045:952-4426,3474,6,0,0,rve,562.SAMN04376786,IS_Tn,1,specI_v3_Cluster95,...,Enterobacterales,Enterobacteriaceae,Escherichia,Escherichia coli,562,SAMN04376786,LVNY01000045:952-4426,LVNY01000045,952,4426
1,1870951.SAMN04252929.LPPF01000080:108-84399,84291,79,0,0,"ser_ce,Phage_integrase,Xer",1870951.SAMN04252929,IS_Tn,1,specI_v3_Cluster77,...,Enterobacterales,Enterobacteriaceae,Enterobacter,Enterobacter roggenkampii,1870951,SAMN04252929,LPPF01000080:108-84399,LPPF01000080,108,84399
2,1313.SAMEA1032643.CNPJ02000077:205410-234256,28846,21,0,0,"DDE_3,DDE_Tnp_IS66",1313.SAMEA1032643,IS_Tn,2,specI_v3_Cluster282,...,Lactobacillales,Streptococcaceae,Streptococcus,Streptococcus pneumoniae,1313,SAMEA1032643,CNPJ02000077:205410-234256,CNPJ02000077,205410,234256
3,1313.SAMEA1032643.CNPJ02000077:179268-184242,4974,3,0,0,huh_y1,1313.SAMEA1032643,IS_Tn,1,specI_v3_Cluster282,...,Lactobacillales,Streptococcaceae,Streptococcus,Streptococcus pneumoniae,1313,SAMEA1032643,CNPJ02000077:179268-184242,CNPJ02000077,179268,184242
4,1313.SAMEA1032643.CNPJ02000077:155048-175733,20685,25,0,0,"Int_Tn916,Phage_integrase,Transposase_mut",1313.SAMEA1032643,IS_Tn,1,specI_v3_Cluster282,...,Lactobacillales,Streptococcaceae,Streptococcus,Streptococcus pneumoniae,1313,SAMEA1032643,CNPJ02000077:155048-175733,CNPJ02000077,155048,175733


In [4]:
print("Number of identified MGE: ", promge_df.mge_genome_position.nunique())
print("Number of analyzed genomes: ", promge_df.genome.nunique())
print("Number of analyzed species: ", promge_df.species.nunique())

Number of identified MGE:  1933280
Number of analyzed genomes:  76060
Number of analyzed species:  7695


## Read proMGE HGT data
- source: https://promge.embl.de/about.cgi 

In [5]:
# read text file into pandas DataFrame
promge_hgt_df = pd.read_csv("dataset/prev_studies/proMGE/HGT_data.txt", sep="\t")

In [6]:
promge_hgt_df.head()

,recombinase_cluster,recombinase,HGT_genome_1,HGT_species_1,HGT_protein_1,MGE_position_1,recombinase_MGE_assignment_1,HGT_genome_2,HGT_species_2,HGT_protein_2,MGE_position_2,recombinase_MGE_assignment_2
0,cas.Cluster_1669,cas1,1204719.PRJEB57,Phocaeicola abscessus,1204719.PRJEB57.GCA_000312445_01601,1204719.PRJEB57.HE997184:324293-334153,IS_Tn,28112.SAMEA4441423,Tannerella forsythia,28112.SAMEA4441423.TFUB20_02022,28112.SAMEA4441423.FMMM01000068:2931-3947,IS_Tn
1,cas.Cluster_1669,cas1,28112.SAMEA4441423,Tannerella forsythia,28112.SAMEA4441423.TFUB20_02022,28112.SAMEA4441423.FMMM01000068:2931-3947,IS_Tn,1204719.PRJEB57,Phocaeicola abscessus,1204719.PRJEB57.GCA_000312445_01601,1204719.PRJEB57.HE997184:324293-334153,IS_Tn
2,cas.Cluster_1717,cas1,726.SAMN03580969,Haemophilus haemolyticus,726.SAMN03580969.AAX18_01790,726.SAMN03580969.LCTK01000006:77242-117737,IS_Tn,484.SAMN03329000,Neisseria flavescens,484.SAMN03329000.TV01_0491,484.SAMN03329000.LAEH01000028:181391-193251,IS_Tn
3,cas.Cluster_1717,cas1,484.SAMN03329000,Neisseria flavescens,484.SAMN03329000.TV01_0491,484.SAMN03329000.LAEH01000028:181391-193251,IS_Tn,726.SAMN03580969,Haemophilus haemolyticus,726.SAMN03580969.AAX18_01790,726.SAMN03580969.LCTK01000006:77242-117737,IS_Tn
4,cas.Cluster_1733,cas1,732.SAMN05327896,Aggregatibacter aphrophilus,732.SAMN05327896.BBB52_07800,732.SAMN05327896.MAQE01000015:50671-63303,IS_Tn,487.SAMN03197529,Neisseria meningitidis,487.SAMN03197529.GCA_001064635_01428,487.SAMN03197529.JVKZ01000042:335-9545,IS_Tn


In [7]:
# compute unique transfer e.g. A<->B is counted once

# ensure consistent ordering within each pair
promge_hgt_df[['m1', 'm2']] = pd.DataFrame(
    promge_hgt_df.apply(lambda x: sorted([x['HGT_genome_1'], x['HGT_genome_2']]), axis=1).tolist(),
    index=promge_hgt_df.index
)

# drop duplicates
unique_df = promge_hgt_df.drop_duplicates(subset=['recombinase_cluster', 'm1', 'm2'])
unique_df['bio_sample_id_1'] = unique_df['HGT_genome_1'].apply(lambda x: x.split('.')[-1])
unique_df['bio_sample_id_2'] = unique_df['HGT_genome_2'].apply(lambda x: x.split('.')[-1])
unique_df['accession_1'] = unique_df['MGE_position_1'].apply(lambda x: x.split('.')[-1].split(':')[0])
unique_df['accession_2'] = unique_df['MGE_position_2'].apply(lambda x: x.split('.')[-1].split(':')[0])
unique_df['HGT_genus_1'] = unique_df['HGT_species_1'].apply(lambda x: x.split(' ')[0])
unique_df['HGT_genus_2'] = unique_df['HGT_species_2'].apply(lambda x: x.split(' ')[0])


In [8]:
## Note: there are some MGEs without the information of their position (ISLAND_ND) -> affect the calculation of the statistics
## ISLAND_ND could indicate site-specific recombinase

# how many recombinase clusters are there?
print("Number of recombinase clusters: ", unique_df.recombinase_cluster.nunique())

# how many recombinase types are there?
print("Number of recombinase types: ", unique_df.recombinase.nunique())

# how many MGE types are there?
mge_hgt_type = set(unique_df['recombinase_MGE_assignment_1']).union(set(unique_df['recombinase_MGE_assignment_2']))
print("Number of MGE types: ", len(mge_hgt_type))

# how many species and genera are there?
hgt_species = set(unique_df['HGT_species_1']).union(set(unique_df['HGT_species_2']))
print("Number of species: ", len(hgt_species))

hgt_genus = set(unique_df['HGT_genus_1']).union(set(unique_df['HGT_genus_2']))
print("Number of genus: ", len(hgt_genus))

#how many unique accession ids in proMGE HGT dataset?
hgt_genomes_1 = unique_df[['HGT_genome_1','MGE_position_1']].drop_duplicates()
hgt_genomes_1.rename(columns={"HGT_genome_1":"HGT_genome","MGE_position_1":"MGE_position"}, inplace=True)
hgt_genomes_2 = unique_df[['HGT_genome_2','MGE_position_2']].drop_duplicates()
hgt_genomes_2.rename(columns={"HGT_genome_2":"HGT_genome","MGE_position_2":"MGE_position"}, inplace=True)
hgt_genomes = pd.concat([hgt_genomes_1, hgt_genomes_2]).drop_duplicates()
#hgt_accession_ids = set(unique_df['HGT_genome_1']).union(set(unique_df['HGT_genome_2']))
print("Number of genomes involved in HGT all: ", hgt_genomes['HGT_genome'].nunique())
hgt_genomes_nd = hgt_genomes[hgt_genomes['MGE_position']=="ISLAND_ND"]
print("Number of genomes involved in HGT with ISlAND_ND label: ", hgt_genomes_nd["HGT_genome"].nunique())

#how many HGT events in proMGE HGT dataset?
hgt_events = unique_df[['HGT_genome_1','HGT_genome_2']].drop_duplicates()
print("Number of HGT events all: ", len(hgt_events))

hgt_events_w_nd = unique_df[(unique_df['recombinase_MGE_assignment_1']=='MGE_ND')|(unique_df['recombinase_MGE_assignment_2']=='MGE_ND')]
print("Number of HGT events including MGE_ND: ", len(hgt_events_w_nd))


n_mge_hgt = len(set(unique_df['MGE_position_1']).union(set(unique_df['MGE_position_2'])))
perc_hgt_no_nd = ((n_mge_hgt-1)/promge_df.mge_genome_position.nunique())*100
print(f"Number of MGEs involved in HGT (excluding MGE_ND): {n_mge_hgt-1} ({perc_hgt_no_nd.__round__(3)}%)")

perc_hgt = (len(hgt_genomes)/promge_df.mge_genome_position.nunique())*100
print(f"Number of MGEs involved in HGT: {len(hgt_genomes)} ({perc_hgt.__round__(3)}%)")

Number of recombinase clusters:  3409
Number of recombinase types:  55
Number of MGE types:  9
Number of species:  2372
Number of genus:  726
Number of genomes involved in HGT all:  4864
Number of genomes involved in HGT with ISlAND_ND label:  848
Number of HGT events all:  13162
Number of HGT events including MGE_ND:  5582
Number of MGEs involved in HGT (excluding MGE_ND): 6183 (0.32%)
Number of MGEs involved in HGT: 7031 (0.364%)


In [9]:
mge_hgt_type

{'CE',
 'Cellular',
 'IS_Tn',
 'Integron',
 'MGE_ND',
 'MI',
 'Phage',
 'Phage_like',
 'nested'}

In [10]:
#unique_df[(unique_df['recombinase_MGE_assignment_1']=='MGE_ND')|(unique_df['recombinase_MGE_assignment_2']=='MGE_ND')]

There are some MGEs involved in HGT, but their position is not defined (ISLAND_ND) 5582 of 13162 HGT events

In [11]:
# check number of genomes per species in the HGT network 
species_1 = unique_df[["HGT_species_1","HGT_genome_1"]]
species_1.rename(columns={"HGT_species_1":"HGT_species","HGT_genome_1":"HGT_genome"}, inplace=True)
species_2 = unique_df[["HGT_species_2","HGT_genome_2"]]
species_2.rename(columns={"HGT_species_2":"HGT_species","HGT_genome_2":"HGT_genome"}, inplace=True)
unique_genome = pd.concat([species_1, species_2]).drop_duplicates()

n_genome_per_species = unique_genome.groupby("HGT_species", as_index=False).agg(
    n_genome=("HGT_genome", "nunique")
).sort_values(by='n_genome',ascending=False)

# Find the 99th percentile threshold
threshold = n_genome_per_species['n_genome'].quantile(0.99)

# Filter the dataframe for rows above that threshold
top_1_percent_df = n_genome_per_species[n_genome_per_species['n_genome'] >= threshold]

top_1_percent_df

,HGT_species,n_genome
1743,Pseudomonas aeruginosa,273
1930,Salmonella enterica,267
552,Clostridioides difficile,92
2069,Staphylococcus aureus,78
2264,Vibrio cholerae,65
2101,Streptococcus agalactiae,52
299,Bacteroides fragilis,46
1238,Listeria monocytogenes,45
2120,Streptococcus pneumoniae,43
2144,Streptococcus suis,40


In [12]:
n_genome_per_species.describe()

,n_genome
count,2372.00000
mean,2.05059
std,8.84140
min,1.00000
25%,1.00000
50%,1.00000
75%,1.00000
max,273.00000


In [13]:
hgt_genomes[hgt_genomes['MGE_position']!="ISLAND_ND"]

,HGT_genome,MGE_position
0,1204719.PRJEB57,1204719.PRJEB57.HE997184:324293-334153
2,726.SAMN03580969,726.SAMN03580969.LCTK01000006:77242-117737
4,732.SAMN05327896,732.SAMN05327896.MAQE01000015:50671-63303
6,1739544.SAMN04480460,1739544.SAMN04480460.KV821652:799-13771
8,1739456.SAMN04480384,1739456.SAMN04480384.KV814812:388020-400695
...,...,...
27154,546.SAMN04124434,546.SAMN04124434.LLJO01000037:223-1179
27156,1796.SAMN03699825,1796.SAMN03699825.LDCO01000054:773-21043
27158,1761786.SAMN04487805,1761786.SAMN04487805.FMAX01000042:14-11616
27160,28108.SAMN03295007,28108.SAMN03295007.JXRW01000006:92970-130964


## Download FASTA files for genomes involved in HGT

In [14]:
# Download FASTA files for genomes involved in HGT according to proMGE database
from Bio import Entrez
from tqdm import tqdm 
import os

Entrez.email = "your.email@example.com"

# genomes involved in HGT
hgt_genomes = hgt_genomes[hgt_genomes['MGE_position']!="ISLAND_ND"]
hgt_genomes['accession'] = hgt_genomes['MGE_position'].apply(lambda x: x.split('.')[-1].split(':')[0])
hgt_accession_ids = hgt_genomes.accession.unique()

fasta_list = []

for acc_id in tqdm(hgt_accession_ids):
    dest_file = f"dataset/genomes/promge/{acc_id}.fasta"

    if not os.path.exists(dest_file):
        try:
            with Entrez.efetch(db="nuccore",
                            id=acc_id,
                            rettype="fasta",
                            retmode="text") as handle:
                fasta_data = handle.read()
                handle.close()
            with open(dest_file, "w") as out:
                out.write(fasta_data)

            #fasta_list.append(fasta_data)
        except Exception as e:
            print(f"Error fetching {acc_id}: {e}")

100%|██████████| 5693/5693 [00:00<00:00, 14719.91it/s]

Error fetching CP016400: HTTP Error 400: Bad Request


## Create proMGE test data set
- negative sampling on genomes involved in HGT
- most MGEs are located in contigs
- take all regions identified as MGEs (regardless of getting involved in HGT) in each genomes, and treat them as "positives"
- then, perform negative sampling on the remaining regions

### Create promge test data set by using core MGEs defined in the database
- core MGEs contain recombinase families of Xer, Cyan, and Arch1
- additional filter: MGEs with mge_category=Cellular, phage_genes=0, and conjugation_genes=0

In [15]:
# get mges involved in HGT
mge_hgt_df = promge_df[promge_df['mge_genome_position'].isin(hgt_genomes.MGE_position.unique())]

# sample some mges not involved in HGT and from species involved in HGT
mge_core = promge_df[(promge_df['mge_category']=='Cellular') & (promge_df['phage_genes']==0) \
       & (promge_df['conjugation_genes']==0) & (promge_df['recombinase'].isin(['Xer','Cyan','Arch1']))]

# get all biosample ids involved in HGT and retrieve the core islands 
# each biosample contains contigs of the same
mge_hgt_biosample_ids = mge_hgt_df.biosample_id.unique()
promge_neg = mge_core[mge_core['biosample_id'].isin(mge_hgt_biosample_ids)][['Accession','Start','End']]
promge_neg['label'] = 0
promge_neg['prediction_method'] = 'promge'

In [16]:
mge_hgt_df[['biosample_id','species']].drop_duplicates().groupby('biosample_id')['species'].nunique().describe()
# each biosample contains exactly 1 species

count    4016.0
mean        1.0
std         0.0
min         1.0
25%         1.0
50%         1.0
75%         1.0
max         1.0
Name: species, dtype: float64

### Create promge test data set with negative sampling

In [39]:
from utils.data_sampling import negatives_sampling

genome_path = 'dataset/genomes/promge'

mge_hgt_samples = mge_hgt_df[['Accession','Start','End']].drop_duplicates()
mge_hgt_samples.columns = [col.lower() for col in mge_hgt_samples.columns]
mge_hgt_samples['prediction_method'] = 'promge'

# Run the negative sampling
promge_dataset = negatives_sampling(mge_hgt_samples, genome_path = genome_path)

In [40]:
promge_dataset

,Accession,Start,End,prediction_method,Label,Sequence,Description,gc
0,JFYY01000019,142,94659,promge,1,TGACGCTGGCGACACTGTTTGACTTTCCCTGGGACGAGCGGGCGCG...,JFYY01000019:142-94659 JFYY01000019.1 Sphingom...,0.622305
1,JULV01000047,75,755,promge,1,TGAATCATTTTAAAGGAAAGCAATTTCAGCAGGATGTGATTATTGT...,JULV01000047:75-755 JULV01000047.1 Corynebacte...,0.372981
2,LSVG01000096,58,1281,promge,1,TGTCACGACGCAAAGAACCTGCCATACCGAATGAGCTTCTCGATCA...,LSVG01000096:58-1281 LSVG01000096.1 Sphingobiu...,0.618464
3,KV816499,140,3055,promge,1,CATCCATTGGTGCCAGTTATCGATGCATTTGGCCGCGAAGGCAGGA...,KV816499:140-3055 KV816499.1 Neisseria sp. HMS...,0.597051
4,LZEX01000042,221,306417,promge,1,TGAACGAACAAAAACTAAAAGACCTTGCAGCAGAATTTGCTAAAGG...,LZEX01000042:221-306417 LZEX01000042.1 Morgane...,0.483104
...,...,...,...,...,...,...,...,...
3618,BCZD01000027,8497,15746,random_neg_sample,0,GGAGATCGGTTGCGGCGTGGGGGCGCACGGGGTCGCCATCGCATCG...,BCZD01000027:8497-15746 BCZD01000027.1 Sphingo...,0.624276
3619,BCZD01000027,2496,11117,random_neg_sample,0,GAGCGATTCCCGCGCGAAGTCGGACTCCAGATTGTCGCGGGGGTCG...,BCZD01000027:2496-11117 BCZD01000027.1 Sphingo...,0.637439
3620,DS996456,766332,786170,random_neg_sample,0,TTGCACAAAGCAATGCAACATATCGAACAAGAAGGTAAGGGAGTAA...,DS996456:766332-786170 DS996456.1 Parabacteroi...,0.435849
3621,CP001655,1931226,1944065,random_neg_sample,0,CATCAATCAACGCATCTATGCGCTGCTGGATCGCACCGACGCCGAC...,CP001655:1931226-1944065 CP001655.1 Dickeya ch...,0.567056


In [41]:
# write the test data to promge_test_neg_sampling.fasta
from utils.util import df_to_fasta_new

# write df to fasta file
param = {'write_file':True,'filename':"dataset/test_folder/promge_test.fasta"} #whether write sequences to a fasta file specified in filename

if not os.path.exists(param['filename']):
    fasta_data = promge_dataset.progress_apply(lambda x: df_to_fasta_new(x,
                                                                dna_only=True, #dna_only determines whether or not to process sequences with IUPAC codes
                                                                query_db=False, #query sequence from database
                                                                **param), axis=1).to_list()
else:
    print(f"{param['filename']} already exists!")

100%|██████████| 9805/9805 [01:43<00:00, 94.35it/s] 


### Read training data pool to determine the taxonomic distance of samples in promge data set

In [43]:
# read who_rki data
who_rki_species = pd.read_excel('dataset/prev_studies/who_rki_sampled_species.xlsx')

In [44]:
def define_test_case(df, training_pool):
    training_species = training_pool.species.unique()
    training_genus = training_pool.genus.unique()
    training_family = training_pool.family.unique()
    training_phylum = training_pool.phylum.unique()

    if df['species'] in training_species:
        return "in_species"
    elif df['species'] not in training_species and df['genus'] in training_genus:
        return "out_species"
    elif df['genus'] not in training_genus and df['family'] in training_family:
        return "out_genera_in_families"
    elif df['genus'] not in training_genus and df['family'] not in training_family and df['phylum'] in training_phylum:
        return "out_genera_out_families"
    else:
        return "out_phylum"

In [46]:
import taxoniq

# map the taxon to the islandviewer_gis_stats DataFrame
def map_taxon_to_df(row):
    try:
        taxon = taxoniq.Taxon(row['taxid'])
        row['scientific_name'] = taxon.scientific_name

        for t in taxon.ranked_lineage:
            row[t.rank.name] = t.scientific_name 
    except Exception as e:
        row['scientific_name'] = None
    return row

promge_tax_df = mge_hgt_df['taxid'].drop_duplicates().reset_index()

# Apply the mapping function in parallel
promge_tax_info_df = promge_tax_df.parallel_apply(map_taxon_to_df, axis=1)
promge_tax_info_df['test_case'] = promge_tax_info_df.parallel_apply(lambda x: define_test_case(x, who_rki_species), axis=1)

In [47]:
tax_test_case_dict = promge_tax_info_df.set_index('taxid')['test_case'].to_dict()
acc_taxid_dict = mge_hgt_df.set_index('Accession')['taxid'].to_dict()

In [48]:
promge_test_df_w_info = pd.merge(promge_dataset, mge_hgt_df[['Accession','phylum','family','genus','species']], 
         on='Accession', how='left').drop_duplicates()

In [49]:
promge_test_df_w_info['taxid'] = promge_test_df_w_info['Accession'].map(acc_taxid_dict)
promge_test_df_w_info['test_case'] = promge_test_df_w_info['taxid'].map(tax_test_case_dict)

In [50]:
promge_test_df_w_info.groupby(['test_case','Label']).size()

test_case                Label
in_species               0         400
                         1        1038
out_genera_in_families   0         207
                         1         358
out_genera_out_families  0        1740
                         1        2860
out_phylum               0         740
                         1         963
out_species              0         536
                         1         963
dtype: int64